### PR TITLE
S2A handshaker client, part 2

### PR DESCRIPTION
**Pull request by matthewstevenson88**
This PR is a follow-up to #21. It contains the implementation of the methods of the `s2a_handshaker_client` that are responsible for calling the S2A handshaker service.

### CODE REVIEW

**Comment by matthewstevenson88** [comment]
This PR is ready for review, once PR #21 is merged. The implementation and the tests are nearly identical to the ALTS counterparts, so it should be pretty easy to review.

**Comment by matthewstevenson88** [comment]
This PR is ready for review!

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc [hunk 1]
```diff
@@ -86,11 +86,14 @@ tsi_result S2AHandshakerClient::ClientStart() {
...
-    gpr_log(GPR_ERROR, kS2AMakeGrpcCallFailed);
+  if (!no_calls_for_testing_) {
+    tsi_result call_result = MakeGrpcCall(/* is_start=*/true);
+    if (call_result != TSI_OK) {
+      gpr_log(GPR_ERROR, kS2AMakeGrpcCallFailed);
+      return call_result;
+    }
   }
-  return call_result;
+  return TSI_OK;
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Is it possible to refactor the code from lines 89 to 94? It seems to be used repeatedly in three different places.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc [hunk 2]
```diff
@@ -175,18 +181,34 @@ tsi_result S2AHandshakerClient::Next(grpc_slice* bytes_received) {
...
+        grpc_slice_to_c_string(client->handshake_status_details());
+    gpr_log(GPR_INFO,
+            "S2AHandshakerClient:%p on_status_received status:%d "
+            "details:|%s| error:|%s|",
+            client, client->handshake_status_code(), status_details,
+            grpc_error_string(error));
+    gpr_free(status_details);
+  }
+  client->MaybeCompleteTsiNext(/* receive_status_finished=*/true,
+                               /* pending_recv_message_result=*/nullptr);
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Don't we also need a global queue (similar to ALTS) to handle a large number of concurrent connection requests?

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h [hunk 3]
```diff
@@ -111,16 +111,43 @@ class S2AHandshakerClient {
    *  no further handshaker requests will be scheduled with the S2A. **/
   void Shutdown();
 
-  /** This method is exposed for testing purposes only. **/
-  grpc_byte_buffer* get_send_buffer_for_testing();
+  void MaybeCompleteTsiNext(
+      bool receive_status_finished,
+      s2a_recv_message_result* pending_recv_message_result);
+
+  /** This method dereferences all of the data owned by this S2A handshaker
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: The comment is confusing a little bit. I think it dereferences a handshaker client and clean up all of its fields once the ref count is zero.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc [hunk 4]
```diff
@@ -235,15 +259,15 @@ tsi_result s2a_handshaker_client_create(
     grpc_pollset_set* interested_parties,
     const grpc_s2a_credentials_options* options, const grpc_slice& target_name,
     grpc_iomgr_cb_func grpc_cb, tsi_handshaker_on_next_done_cb cb,
-    void* user_data, bool is_client, S2AHandshakerClient** client) {
+    void* user_data, bool is_client, bool is_test, S2AHandshakerClient** client) {
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Please also change the function name to adhere to C++ naming convention.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc [hunk 5]
```diff
@@ -259,14 +283,16 @@ static void S2AHandshakerCallUnref(void* arg, grpc_error* error) {
...
+void S2AHandshakerClient::Unref() {
+  if (gpr_unref(refs_)) {
     if (call_ != nullptr) {
       grpc_core::ExecCtx::Run(DEBUG_LOCATION,
                               GRPC_CLOSURE_CREATE(S2AHandshakerCallUnref, call_,
                                                   grpc_schedule_on_exec_ctx),
                               GRPC_ERROR_NONE);
     }
+    gpr_free(refs_);
+    refs_ = nullptr;
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Why do we set it `nullptr` after freeing it?

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc [hunk 6]
```diff
@@ -275,11 +301,14 @@ S2AHandshakerClient::~S2AHandshakerClient() {
     grpc_slice_unref_internal(recv_bytes_);
     grpc_slice_unref_internal(target_name_);
     gpr_free(buffer_);
+    buffer_ = nullptr;
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Same as above.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h [hunk 7]
```diff
@@ -186,10 +213,17 @@ class S2AHandshakerClient {
...
    *  op is done. **/
-  bool receive_status_finished_;
+  bool receive_status_finished_ = false;
   /** If this field is not nullptr, then it contains arguments needed to
    *  complete a TSI next callback. **/
-  s2a_recv_message_result* pending_recv_message_result_;
+  s2a_recv_message_result* pending_recv_message_result_ = nullptr;
+  /** This variable should be set to true iff the S2A handshaker client instance
+   *  is instantiated as part of a test. **/
+  bool is_test_ = false;
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: Can we consolidate these two variables? `is_test_` and `no_calls_for_testing_`

**Comment by yihuazhang** [comment]
Sorry for the delay Matt. 

My major concern is I am not convinced about the correctness of some changes in this PR without seeing more comprehensive tests:   

1: Apolcyn@ made a series of changes to ALTS code to make it resistant against a large number of concurrent connection requests (https://github.com/grpc/grpc/pull/20722, https://github.com/grpc/grpc/pull/21189, https://github.com/grpc/grpc/pull/20929). In this PR, I noticed a similar but a little different changes are made (e.g., I did not see the use of connection queues) so not sure if the changes indeed work without doing more integration tests (similar to `alts_concurrent_connectivity_test.cc`). Basically, we did not verify any interaction to the handshake server. 

2: Do you think it is necessary to do some code refactoring between ALTS and S2A? If the long term plan is to move the S2A code out of gRPC repo, it may not be necessary though.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc (hunk 1 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Done in next commit.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc (hunk 2 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: As we discussed offline, we do not need the global queue. Limiting the number of concurrent connection requests will be handled by the S2A service.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc (hunk 4 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Done in next commit (and same for the destroy method).

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc (hunk 5 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: It is not needed. Good catch! I've removed it in the next commit.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc (hunk 6 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Also removed it.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 3 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: I've made this comment more precise in the next commit. PTAL.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 7 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: I kept them separate for the following reason: there are basic unit tests in `s2a_handshaker_client_test.cc`, and then unit tests for the methods that call the S2A service in `s2a_handshaker_client_caller_test.cc`. 

The variable `is_test_` is set to `true` in both cases (and in certain TSI handshaker tests). The variable `no_calls_for_testing_` is set to `true` in the former case, and `false` in the latter case.

Note that only `is_test_` must be specified in the constructor of the `S2AHandshakerClient`.

What are your thoughts on this strategy?

**Comment by matthewstevenson88** [comment]
Thanks for the comments @yihuazhang! I really appreciate them. I've addressed them all in the most recent commit, or responded to them directly.

Below, I will respond to your 2 comments above. (Some of this is repetition from our offline discussion.)

1. As we discussed offline, our team has agreed that the S2A service will be responsible for capping the number of concurrent connection requests. As a consequence, the global queue is not necessary. (There is an email chain with apolcyn@, on which you are cc-ed, that verifies what portions of code will be removed from ALTS along with the global queue. I've removed them in the S2A client already.)

As for the additional tests, I propose the following: when two more PR's are merged (this one and one other yet to be posted), there will be a complete TSI handshaker and handshaker client implementation. Once this is done, I will post a PR that contains the following additional tests:

- The S2A version of the fake handshaker service, with a cap on the number of concurrent connection requests.
- The S2A version of the tests in `alts_tsi_handshaker_test.cc` (i.e. the ones that use the fake handshaker service to mock a handshake).
- The S2A version of the `alts_concurrent_connectivity_test.cc`.

How does that plan look to you?

2. What code refactoring are you referring to? Once the S2A client is ultimately merged somewhere, I will do some reorganization of the ALTS code, namely putting some stuff in a shared library under `src/core/tsi`. Is this what you are referring to?

In terms of the long term plan, if we decide to move the S2A client out of gRPC, there will also be refactoring to avoid the use of gRPC-internal API's.

**Comment by yihuazhang** [comment]
The plan SGTM. Regarding refactoring since it seems S2A will not be placed in OSS gRPC repo, let's do not do any refactoring regarding code sharing between ALTS and S2A.

**Comment by matthewstevenson88** [comment]
Per our offline discussion, we will create an S2A-dedicated copy of the `gsec_aead_crypter`, to avoid any refactoring of the ALTS code.

Granted the plan, is this PR good to go? Or did you want to take another look at it?

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h [hunk 8]
```diff
@@ -111,16 +111,47 @@ class S2AHandshakerClient {
...
+  void set_no_calls_for_testing(bool no_calls);
 
  private:
   /** This method makes a call to the S2A service. **/
   tsi_result MakeGrpcCall(bool is_start);
 
-  void MaybeCompleteTsiNext(
-      bool receive_status_finished,
-      s2a_recv_message_result* pending_recv_message_result);
+  tsi_result MakeGrpcCallUtil(bool is_start);
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: Add some comments for it?

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc (hunk 2 above)
Comment (yihuazhang) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Ack.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h [hunk 9]
```diff
@@ -111,16 +111,47 @@ class S2AHandshakerClient {
...
+
+  /** These methods are exposed for use in the |on_status_received| callback
+   *  function. **/
+  grpc_status_code handshake_status_code() { return handshake_status_code_; }
+  const grpc_slice handshake_status_details() {
+    return handshake_status_details_;
+  }
+
+  /** These methods are exposed for testing purposes only. If |is_test_| is
+   *  false, then these methods do nothing or return nullptr. **/
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: `and`?

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc [hunk 10]
```diff
@@ -289,8 +305,49 @@ void s2a_handshaker_client_destroy(S2AHandshakerClient* client) {
...
 
-grpc_byte_buffer* S2AHandshakerClient::get_send_buffer_for_testing() {
-  return send_buffer_;
+void S2AHandshakerClient::set_grpc_caller_for_testing(s2a_grpc_caller caller) {
+  if (is_test_) {
+    grpc_caller_ = caller;
+  }
+}
+
+grpc_metadata_array* S2AHandshakerClient::initial_metadata_for_testing() {
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: The logic here is a little confusing. I expected that if `is_test_` is true, `nullptr` should be returned which does not seem to be the case.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc [hunk 11]
```diff
@@ -30,16 +32,208 @@ namespace experimental {
...
+  GPR_ASSERT(grpc_caller_ != nullptr);
+  if (grpc_caller_(call_, ops, static_cast<size_t>(op - ops),
+                   &on_handshaker_service_resp_recv_) != GRPC_CALL_OK) {
+    gpr_log(GPR_ERROR, "Start batch operation failed");
+    return TSI_INTERNAL_ERROR;
+  }
+  return TSI_OK;
 }
 
+/** The implementation of this method is nearly identical to its ALTS
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc]: Not sure if we want to keep this comment especially if we do not want to place the code in gRPC repo.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc [hunk 12]
```diff
@@ -30,16 +32,208 @@ namespace experimental {
  *  function of the |S2AHandshakerClient| class. This method enables the S2A
  *  handshaker client to make a gRPC call to the S2A service. **/
 
+const size_t kHandshakerClientOpNum = 4;
+
+/** The implementation of |make_grpc_call| is nearly identical to that of its
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc]: Similar to the comment below.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 8 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: Done in next commit.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 7 above)
Comment (yihuazhang) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: IIUC, `MakeGrpcCall` is not actually called in both cases since `grpc_caller_` callback (which talks to the handshaker server) is not called so not sure why we treat those two tests differently.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h [hunk 13]
```diff
@@ -111,16 +111,47 @@ class S2AHandshakerClient {
    *  no further handshaker requests will be scheduled with the S2A. **/
   void Shutdown();
 
-  /** This method is exposed for testing purposes only. **/
-  grpc_byte_buffer* get_send_buffer_for_testing();
+  void MaybeCompleteTsiNext(
```
Comment (yihuazhang) [inline src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: Can we add some comments for it? especially why it is `Maybe`?

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 9 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: Sorry, I meant: if the return type is `void`, do nothing; if the return type is a pointer, then do nothing and return `nullptr`. I've clarified this comment in the next commit.

**Comment by yihuazhang** [comment]
Thanks for the changes Matt. My major concern is that since we do not want to place the code inside gRPC repo (either a short or long term), I think the comments in the code should not or at least minimize its reference to `ALTS`. Please add more comments to some of the API's (`MaybeCompleteTsiNext`) to make it easier to understand in future. You can find some conversations I had with Alex in https://github.com/grpc/grpc/pull/20929/files about it.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc (hunk 10 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Since this is a `for_testing` method, it should only return something interesting if `is_test_` is true. If `is_test_` is false, then it should return `nullptr`. (The confusion here might be related to the previous comment, about the comments on the `for_testing` functions that were unclear.)

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc (hunk 11 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc]: Removed in next commit.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc (hunk 12 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc]: Removed in next commit.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 13 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: I've added some comments in the next commit.

**Comment by matthewstevenson88** [comment]
Thanks for the comments @yihuazhang! I've responded to them all. In particular, I've removed all references to `ALTS` and added more comments (e.g. to the `MaybeCompleteTsiNext` API).

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 7 above)
Comment (yihuazhang) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: I did not see this comment is addressed.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc (hunk 10 above)
Comment (yihuazhang) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: I see. Please also add above comments in the code to make it clear.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 13 above)
Comment (yihuazhang) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: I do not think the comment is accurate. 

If the handshaker client has a final result or failure, it checks if the status is received or not, and if it is received then it will invoke the TSI callback and if the status is not received, it will not invoke the callback. If the final result is not received, it will go ahead execute the callback.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 7 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: Sorry Yihua I missed this comment on my last pass. The issue is the following: for the tests in `s2a_handshaker_client_test.cc`, there is no `grpc_caller_` set and, if the `no_calls_for_testing_` variable is not present in `MakeGrpcCallUtil, then `MakeGrpcCall` (and hence `grpc_caller_`) is called. This leads to an error in the test.

After looking at this more carefully, I think it makes sense to do the following:
1. Remove the `s2a_handshaker_client_test.cc` file entirely and add any edge cases that it tests to `s2a_handshaker_client_caller_test.cc`.
2. Remove the `no_calls_for_testing_` variable.

WDYT?

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc (hunk 10 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc]: Clarified this comment in the next commit.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 7 above)
Comment (yihuazhang) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: The plan (removing the boolean flag and combining these two tests) SGTM.

Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: Sounds good. Made these changes in the next commit.

File: src/core/tsi/s2a/handshaker/s2a_handshaker_client.h (hunk 13 above)
Comment (matthewstevenson88) [reply src/core/tsi/s2a/handshaker/s2a_handshaker_client.h]: I've corrected this in the newest commit. Thanks! :)

**Comment by matthewstevenson88** [comment]
Thanks for all the comments @yihuazhang! I've implemented the changes we discussed in the most recent commit.

**Comment by matthewstevenson88** [comment]
Thanks for the review @yihuazhang!

Squashing and merging now.
